smarts/core/plan.py

```diff
@@ -0,0 +1,337 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# to allow for typing to refer to class being defined (Mission)...
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from functools import lru_cache
+import math
+import random
+from typing import List, Optional, Tuple
+
+import numpy as np
+
+from .coordinates import Dimensions, Heading, Point, Pose, RefLinePoint
+from .road_map import RoadMap
+
+from smarts.sstudio.types import EntryTactic, TrapEntryTactic
+
+
+class PlanningError(Exception):
+    pass
+
+
+# XXX: consider using smarts.core.coordinates.Pose for this
+@dataclass(frozen=True)
+class Start:
+    position: Tuple[int, int]
+    heading: Heading
+    from_front_bumper: Optional[bool] = True
+
+    @property
+    def point(self) -> Point:
+        return Point(*self.position)
+
+    @classmethod
+    def from_pose(cls, pose: Pose):
+        return cls(
+            position=pose.position[:2],
+            heading=pose.heading,
+            from_front_bumper=False,
+        )
+
+
+@dataclass(frozen=True)
+class Goal:
+    def is_endless(self) -> bool:
+        return True
+
+    def is_reached(self, vehicle) -> bool:
+        return False
+
+
+@dataclass(frozen=True)
+class EndlessGoal(Goal):
+    pass
+
+
+@dataclass(frozen=True)
+class PositionalGoal(Goal):
+    position: Point
+    # target_heading: Heading
+    radius: float
+
+    @classmethod
+    def from_road(
+        cls,
+        road_id: str,
+        road_map: RoadMap,
+        lane_index: int = 0,
+        lane_offset: float = None,
+        radius: float = 1,
+    ):
+        road = road_map.road_by_id(road_id)
+        lane = road.lane_at_index(lane_index)
+
+        if lane_offset is None:
+            # Default to the midpoint safely ensuring we are on the lane and not
+            # bordering another
+            lane_offset = lane.length * 0.5
+
+        position = lane.from_lane_coord(RefLinePoint(lane_offset))
+        return cls(position=position, radius=radius)
+
+    def is_endless(self) -> bool:
+        return False
+
+    def is_reached(self, vehicle) -> bool:
+        a = vehicle.position
+        b = self.position
+        sqr_dist = (a[0] - b.x) ** 2 + (a[1] - b.y) ** 2
+        return sqr_dist <= self.radius ** 2
+
+
+@dataclass(frozen=True)
+class TraverseGoal(Goal):
+    """A TraverseGoal is satisfied whenever an Agent-driven vehicle
+    successfully finishes traversing a non-closed (acyclical) map
+    It's a way for the vehicle to exit the simulation successfully,
+    for example, driving across from one side to the other on a
+    straight road and then continuing off the map.  This goal is
+    non-specific about *where* the map is exited, save for that
+    the vehicle must be going the correct direction in its lane
+    just prior to doing so."""
+
+    def __init__(self, road_map: RoadMap):
+        super().__init__()
+        self._road_map = road_map
```

Comment:
I believe assigning to a field here will throw an exception if `frozen=True`. This issue came up on another branch (see https://github.com/huawei-noah/SMARTS/pull/1060#discussion_r690406868). Removing the frozen param will cause an error in Python 3.8 (see https://github.com/huawei-noah/SMARTS/pull/1029). If we want to keep everything frozen, while avoiding storing `road_map` on this object, perhaps we could use an approach like in `PositionalGoal` above (might require storing a closure). Or we could remove the `dataclass` decorator as suggested in the linked discussion.

Reply:
Good catch!  Yes, you are right.  For now, I just pushed an update such that `TraverseGoal` (and the `Goal` base class) are not frozen.  But as Tucker implied in the comment you linked, we probably need to give more thought to this long term.

However, I also note that the rest of `TraverseGoal` currently won't yet work for the `RoadMap` api either, as we do not support accessing Sumo "nodes" via the api.  This is something we're going to have to address in a follow-on PR.

Reply:
actually, thanks to this, I was just inspired to look again at the `_drove_off_map()` method in `TraverseGoal` and it wasn't that hard to implement, so I just went ahead and did it and pushed it.  Thanks again for bringing this up!